### PR TITLE
Add additional tests for mom dynamic resources

### PR DESCRIPTION
**Pull request by dinesh1306**
#### Bug/feature Description
* Add automated tests to verify mom dynamic resource.

#### Affected Platform(s)
* All

#### Solution Description
 Added tests for mom dynamic resource of following scenario.
* Host level string resource with mom dynamic value string.
* Host level consumable float resource with mom dynamic value float.
* Multiple host level resource with all correct values(return same type value from host script). 
* Delete a resource while jobs are running/queued.
* Alter no. of resource value in host script.
* multiple host level resource one or more incorrect values(return different type value from host script).

#### Checklist:

- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
   - [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [ ] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
   - [ ] I have added  **manual test(s) to this pull request and explained why PTL is not appropriate** for this case.
- [ ] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to this pull request** as evidence of testing/verification.

__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

### CODE REVIEW

**Comment by dinesh1306** [comment]
[res_TestMomDynRes.txt](https://github.com/PBSPro/pbspro/files/1921281/res_TestMomDynRes.txt)
execution result.

File: test/tests/functional/pbs_mom_dynamic_resource_ext.py [hunk 1]
```diff
@@ -60,6 +60,36 @@ def create_mom_resources(self, resc_name,
...
+                                  resc_type, resc_flag,
+                                  script_body):
+        """
+        helper function for create multiple mom_resources.
+        """
+        fp_list = []
+
+        resc_name_mom = ",".join(resc_name)
+        resc_name_mom = '"' + resc_name_mom + '"'
+        resc_name_svr = ",".join(resc_name)
```
Comment (anamikau) [inline test/tests/functional/pbs_mom_dynamic_resource_ext.py]: why not assigning resc_name_svr = resc_name_mom before line 74

File: test/tests/functional/pbs_mom_dynamic_resource_ext.py [hunk 2]
```diff
@@ -60,6 +60,36 @@ def create_mom_resources(self, resc_name,
...
+        self.scheduler.set_sched_config({'mom_resources': resc_name_mom},
+                                        validate=True)
+        self.scheduler.add_resource(resc_name_svr)
+
+        for i in range(2):
```
Comment (anamikau) [inline test/tests/functional/pbs_mom_dynamic_resource_ext.py]: can you explain why you are looping 2 times

File: test/tests/functional/pbs_mom_dynamic_resource_ext.py [hunk 3]
```diff
@@ -60,6 +60,36 @@ def create_mom_resources(self, resc_name,
                                         validate=True)
         self.scheduler.add_resource(resc_name)
 
+        return fp
+
+    def create_mult_mom_resources(self, resc_name,
+                                  resc_type, resc_flag,
+                                  script_body):
+        """
+        helper function for create multiple mom_resources.
```
Comment (anamikau) [inline test/tests/functional/pbs_mom_dynamic_resource_ext.py]: helper function for creating

Comment (anamikau) [reply test/tests/functional/pbs_mom_dynamic_resource_ext.py]: Also can we combine these 2 functions related to adding mom resources. Just a suggestion but if it adds too much complexity then this is also okay.

File: test/tests/functional/pbs_mom_dynamic_resource_ext.py [hunk 4]
```diff
@@ -145,3 +175,177 @@ def test_t6(self):
...
+        """
+        host level consumable float resource with mom dynamic value float
+        """
+        resc_name = "foo"
+        self.create_mom_resources(resc_name, "float", "nh", "/bin/echo 3")
+
+        a = {'scheduling': 'True'}
+        self.server.manager(MGR_CMD_SET, SERVER, a)
+
+        # Submit a job that requests mom dynamic resource
```
Comment (anamikau) [inline test/tests/functional/pbs_mom_dynamic_resource_ext.py]: update the comment saying "Submit a job that requests more value for the mom dynamic resource than available"

File: test/tests/functional/pbs_mom_dynamic_resource_ext.py [hunk 5]
```diff
@@ -145,3 +175,177 @@ def test_t6(self):
...
+        j = Job(TEST_USER, attrs=attr)
+        jid = self.server.submit(j)
+
+        # The job shouldn't run
+        c = "Can Never Run:"
+        c += " Insufficient amount of resource: foo (R: 4 A: 3 T: 3)"
+        self.server.expect(JOB, {'job_state': 'Q', 'comment': c},
+                           id=jid, attrop=PTL_AND)
+
+        # Submit a job that requests mom dynamic resource
```
Comment (anamikau) [inline test/tests/functional/pbs_mom_dynamic_resource_ext.py]: update the comment

File: test/tests/functional/pbs_mom_dynamic_resource_ext.py [hunk 6]
```diff
@@ -145,3 +175,177 @@ def test_t6(self):
...
+        attr = {"Resource_List." + resc_name[0]: 2,
+                "Resource_List." + resc_name[1]: 3}
+        j = Job(TEST_USER, attrs=attr)
+        jid = self.server.submit(j)
+
+        # The job should run
+        self.server.expect(JOB, {'job_state': 'R'}, id=jid)
+
+        # Submit a job that requests no. mom resource
+        # greater than scrtipt value
```
Comment (anamikau) [inline test/tests/functional/pbs_mom_dynamic_resource_ext.py]: update the comment as "Submit a job that requests more value than available for both resources"

File: test/tests/functional/pbs_mom_dynamic_resource_ext.py [hunk 7]
```diff
@@ -145,3 +175,177 @@ def test_t6(self):
...
+
+    def test_t10(self):
+        """
+        multiple host level resource with one or more incorrect values.
+        Submit different jobs with different resource types
+        """
+
+        resc_name = ["foo", "foo1"]
+        resc_type = ["float", "string"]
+        script_body = ["/bin/echo 3", "/bin/echo 3.14"]
```
Comment (anamikau) [inline test/tests/functional/pbs_mom_dynamic_resource_ext.py]: like we have learnt "3.14" will be treated as string and not float. I suggest changing this resource to size and pass string or float.

File: test/tests/functional/pbs_mom_dynamic_resource_ext.py [hunk 8]
```diff
@@ -145,3 +175,177 @@ def test_t6(self):
...
+        self.server.expect(JOB, {'job_state': 'R'}, id=jid)
+
+        try:
+            self.server.manager(MGR_CMD_DELETE, RSC, id='foo')
+        except PbsManagerError as e:
+            self.assertTrue("Resource busy on server")
+
+    def test_t12(self):
+        """
+        Alter no. of mom dynamic resource will change Job state.
```
Comment (anamikau) [inline test/tests/functional/pbs_mom_dynamic_resource_ext.py]: change no. to number

File: test/tests/functional/pbs_mom_dynamic_resource_ext.py [hunk 9]
```diff
@@ -145,3 +175,177 @@ def test_t6(self):
...
+
+        with open(filename, "wb") as fd:
+            fd.truncate()
+            fd.write("echo 1")
+
+        a = {'scheduling': 'True'}
+        self.server.manager(MGR_CMD_SET, SERVER, a)
+
+        self.server.rerunjob(jobid=jid)
+        # The job should run
```
Comment (anamikau) [inline test/tests/functional/pbs_mom_dynamic_resource_ext.py]: update the comment to job will not run

**Comment by dinesh1306** [comment]
@anamikau , i have addressed your comments. Please take look.

File: test/tests/functional/pbs_mom_dynamic_resource_ext.py (hunk 2 above)
Comment (dinesh1306) [reply test/tests/functional/pbs_mom_dynamic_resource_ext.py]: number of elements in the list are 2 so i am iterating 2 times. now i am first caluculate the length of list and then iterate that much time.

File: test/tests/functional/pbs_mom_dynamic_resource_ext.py (hunk 3 above)
Comment (dinesh1306) [reply test/tests/functional/pbs_mom_dynamic_resource_ext.py]: combining 2 functions is complicated, because for function  "create_mult_mom_resources" list are  passed as a parameter and for function "create_mom_resources" only values are passed as parameter. if we want to combine 2 functions then for single value also we need to create list and then that list pass to the function as a parameter.

File: test/tests/functional/pbs_mom_dynamic_resource_ext.py (hunk 7 above)
Comment (dinesh1306) [reply test/tests/functional/pbs_mom_dynamic_resource_ext.py]: if i change resource to size and pass string in the host script then it is same scenario as PP-1263. 
I tried with creating size resource and pass float in host script at that time job comes in 'R' state because 3.14 can be used for size.
After that i am passing negative value in host script and ask positive value in job sumbission, then i check job state and job comment .

File: test/tests/functional/pbs_mom_dynamic_resource_ext.py [hunk 10]
```diff
@@ -316,11 +314,11 @@ def test_t11(self):
         try:
             self.server.manager(MGR_CMD_DELETE, RSC, id='foo')
         except PbsManagerError as e:
-            self.assertTrue("Resource busy on server")
+            self.logger.info(e.message)
```
Comment (anamikau) [inline test/tests/functional/pbs_mom_dynamic_resource_ext.py]: why did you remove the assertion? I think you need to update it to something
assertTrue("Resource busy on server" in e.msg[0])

File: test/tests/functional/pbs_mom_dynamic_resource_ext.py [hunk 11]
```diff
@@ -145,3 +172,178 @@ def test_t6(self):
...
+        """
+        Test for host level string resource
+        with string mom dynamic value
+        """
+
+        resc_name = "foo"
+        self.create_mom_resources(resc_name, "string", "h", "/bin/echo abc")
+
+        a = {'scheduling': 'True'}
+        self.server.manager(MGR_CMD_SET, SERVER, a)
```
Comment (bhroam) [inline test/tests/functional/pbs_mom_dynamic_resource_ext.py]: What does this accomplish?  There should be no jobs to run.

File: test/tests/functional/pbs_mom_dynamic_resource_ext.py [hunk 12]
```diff
@@ -145,3 +172,178 @@ def test_t6(self):
...
+        resc_name = ["foo", "foo1"]
+        resc_type = ["float", "long"]
+        script_body = ["/bin/echo 3", "/bin/echo 4"]
+        resc_flag = ["nh", "nh"]
+
+        self.create_mult_mom_resources(resc_name, resc_type,
+                                       resc_flag, script_body)
+
+        a = {'scheduling': 'True'}
+        self.server.manager(MGR_CMD_SET, SERVER, a)
```
Comment (bhroam) [inline test/tests/functional/pbs_mom_dynamic_resource_ext.py]: As above, what does this accomplish?

File: test/tests/functional/pbs_mom_dynamic_resource_ext.py [hunk 13]
```diff
@@ -145,3 +172,178 @@ def test_t6(self):
...
+        resc_name = ["foo", "foo1"]
+        resc_type = ["float", "size"]
+        script_body = ["/bin/echo 3", "/bin/echo -2"]
+        resc_flag = ["nh", "nh"]
+
+        self.create_mult_mom_resources(resc_name, resc_type,
+                                       resc_flag, script_body)
+
+        a = {'scheduling': 'True'}
+        self.server.manager(MGR_CMD_SET, SERVER, a)
```
Comment (bhroam) [inline test/tests/functional/pbs_mom_dynamic_resource_ext.py]: As above, what does this accomplish (and for the rest of the test suite).

File: test/tests/functional/pbs_mom_dynamic_resource_ext.py [hunk 14]
```diff
@@ -145,3 +172,178 @@ def test_t6(self):
...
+        j = Job(TEST_USER, attrs=attr)
+        jid = self.server.submit(j)
+
+        # The job should run
+        self.server.expect(JOB, {'job_state': 'R'}, id=jid)
+
+        try:
+            self.server.manager(MGR_CMD_DELETE, RSC, id='foo')
+        except PbsManagerError as e:
+            self.logger.info(e.message)
```
Comment (bhroam) [inline test/tests/functional/pbs_mom_dynamic_resource_ext.py]: This will be confusing.  You are logging an error message, but passing your test.  Either just 'pass' here or log "Correctly unable to delete busy resource"

File: test/tests/functional/pbs_mom_dynamic_resource_ext.py [hunk 15]
```diff
@@ -145,3 +172,178 @@ def test_t6(self):
...
+        attr = {"Resource_List." + resc_name: '2'}
+        j = Job(TEST_USER, attrs=attr)
+        jid = self.server.submit(j)
+
+        # The job should run
+        self.server.expect(JOB, {'job_state': 'R'}, id=jid)
+
+        with open(filename, "wb") as fd:
+            fd.truncate()
+            fd.write("echo 1")
```
Comment (bhroam) [inline test/tests/functional/pbs_mom_dynamic_resource_ext.py]: I just realized something that will affect your entire test suite.  You are not deleting the mom dynamic resource scripts.  I'd modify your create_mom_resource* functions to store the filenames into a list on the class.  You can then remove them in tearDown().

File: test/tests/functional/pbs_mom_dynamic_resource_ext.py [hunk 16]
```diff
@@ -60,6 +60,33 @@ def create_mom_resources(self, resc_name,
                                         validate=True)
         self.scheduler.add_resource(resc_name)
 
+        return fp
+
+    def create_mult_mom_resources(self, resc_name,
+                                  resc_type, resc_flag,
+                                  script_body):
+        """
```
Comment (bhroam) [inline test/tests/functional/pbs_mom_dynamic_resource_ext.py]: I'd merge the two create functions as @anamikau suggested.  I don't think it is hard to pass a list of one item instead of the item itself.  If you do want to keep the functions separate, at least have this function call the first one.

File: test/tests/functional/pbs_mom_dynamic_resource.py [hunk 17]
```diff
@@ -264,7 +251,7 @@ def test_t9(self):
         self.server.expect(JOB, {'job_state': 'Q', 'comment': c},
                            id=jid, attrop=PTL_AND)
 
-    def test_t10(self):
+    def test_mulptiple_res_invalid_value(self):
```
Comment (anamikau) [inline test/tests/functional/pbs_mom_dynamic_resource.py]: typo multiple

File: test/tests/functional/pbs_mom_dynamic_resource.py [hunk 18]
```diff
@@ -293,57 +277,67 @@ def test_t10(self):
...
 
         # The job should run
         self.server.expect(JOB, {'job_state': 'R'}, id=jid)
 
+        # Check for the expected error message throwing by PBS
         try:
             self.server.manager(MGR_CMD_DELETE, RSC, id='foo')
         except PbsManagerError as e:
-            self.logger.info(e.message)
+            self.logger.info("As expected got error message: " + e.message)
```
Comment (anamikau) [inline test/tests/functional/pbs_mom_dynamic_resource.py]: you are not checking for anything but just logging, update this to assertTrue (message in e.message)

File: test/tests/functional/pbs_mom_dynamic_resource.py [hunk 19]
```diff
@@ -293,57 +277,67 @@ def test_t10(self):
...
 
-    def test_t12(self):
+    def test_alter_res_value(self):
         """
         Alter number of mom dynamic resource will change Job state.
```
Comment (anamikau) [inline test/tests/functional/pbs_mom_dynamic_resource.py]: update docstring as "Verify that altered mom dynamic resource value will take affect only when a job is rerun"

**Comment by dinesh1306** [comment]
@anamikau , I have incorporate your comments, Please take a look.

File: test/tests/functional/pbs_mom_dynamic_resource.py [hunk 20]
```diff
@@ -282,6 +282,7 @@ def test_delete_res(self):
         delete a resource while jobs are running/queued.
         """
 
+        message = "qmgr obj=foo svr=default: Resource busy on job"
```
Comment (anamikau) [inline test/tests/functional/pbs_mom_dynamic_resource.py]: please remove  "qmgr obj=foo svr=default:" from the error message

Comment (bhroam) [reply test/tests/functional/pbs_mom_dynamic_resource.py]: I agree.  That will allow the test to pass in API mode.

**Comment by dinesh1306** [comment]
@anamikau  i have incorporate your comments , please take look.

**Comment by dinesh1306** [comment]
final execution result 
[res.txt](https://github.com/PBSPro/pbspro/files/1940987/res.txt)

**Comment by dinesh1306** [comment]
squashed and rebased , ready for merge.
